### PR TITLE
feat: HIP-904 Introduce `unlimitedAutoAssociations` feature flag

### LINKED ISSUE #12981: HIP-904: Add feature flag for unlimited token associations. Modify TokenConfig and EntityConfig to ensure we are not limiting maxAutoAssociations.

### PR DESCRIPTION
**Pull request by stoyanov-st**
**Description**:
With this PR we are introducing the `unlimitedAutoAssociations` feature flag in `EntitiesConfig`.
We also add relative changes to handler logic in regards to auto associations and relative tests.
**Related issue(s)**:

Fixes #12981 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by MrValioBg** [comment]
It seems like this PR includes adding the feature flag for both the mono and modularized codebase. Why do we need it in mono?

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAssociateToAccountHandler.java [hunk 1]
```diff
@@ -170,18 +180,22 @@ private boolean isTotalNumTokenRelsWithinMax(
...
+            @NonNull final ContractsConfig contractsConfig,
             @NonNull final Account account,
             @NonNull final List<TokenID> tokenIds) {
         final var numAssociations = requireNonNull(account).numberAssociations();
         final var tokenAssociationsLimited = entitiesConfig.limitTokenAssociations();
         final var maxTokensPerAccount = config.maxPerAccount();
-        return !tokenAssociationsLimited || (numAssociations + tokenIds.size() <= maxTokensPerAccount);
+        return !tokenAssociationsLimited
+                || ((numAssociations + tokenIds.size() <= maxTokensPerAccount)
+                        && contractsConfig.unlimitedAutoAssociations());
```
Comment (MrValioBg) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAssociateToAccountHandler.java:198]: Shouldn't this be 
`return !tokenAssociationsLimited || contractsConfig.unlimitedAutoAssociations()  ||
            (numAssociations + tokenIds.size() <= maxTokensPerAccount);`
So its || instead of &&

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java [hunk 2]
```diff
@@ -83,4 +83,5 @@ public record ContractsConfig(
                 boolean chargeGasOnPreEvmException,
         @ConfigProperty(value = "evm.nonExtantContractsFail", defaultValue = "0") @NetworkProperty
                 Set<Long> evmNonExtantContractsFail,
-        @ConfigProperty(value = "evm.version", defaultValue = "v0.50") @NetworkProperty String evmVersion) {}
+        @ConfigProperty(value = "evm.version", defaultValue = "v0.50") @NetworkProperty String evmVersion,
+        @ConfigProperty(defaultValue = "true") @NetworkProperty boolean unlimitedAutoAssociations) {}
```
[the remark this answers is not in the document]
Comment (stoyanov-st) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java:87]: Mainly because we have the `allowAutoAssociations` in the same config.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java [hunk 3]
```diff
@@ -183,12 +184,14 @@ private void validateSupplyType(final TokenSupplyType supplyType, final long max
...
+            @NonNull final ContractsConfig contractsConfig,
             @NonNull final Account account,
             @NonNull final Token token,
             @NonNull final WritableTokenRelationStore tokenRelStore) {
         validateFalse(
-                entitiesConfig.limitTokenAssociations()
-                        && account.numberAssociations() + 1 > tokensConfig.maxPerAccount(),
+                (entitiesConfig.limitTokenAssociations()
+                                && account.numberAssociations() + 1 > tokensConfig.maxPerAccount())
+                        || !contractsConfig.unlimitedAutoAssociations(),
```
Comment (MrValioBg) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java:194]: Do we need the unlimitedAutoAssociations, here? Aren't we overriding the limitTokenAssociations, one?

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAssociateToAccountHandler.java (hunk 1 above)
Comment (MrValioBg) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAssociateToAccountHandler.java]: Discussed offline, we shouldn't override the tokenAssociationsLimited flag with the feature toggle.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java [hunk 4]
```diff
@@ -191,8 +191,8 @@ public void validateAssociation(
             @NonNull final Token token,
             @NonNull final WritableTokenRelationStore tokenRelStore) {
         validateFalse(
-                entitiesConfig.limitTokenAssociations()
-                        && account.numberAssociations() + 1 > tokensConfig.maxPerAccount(),
+                (entitiesConfig.limitTokenAssociations()
+                        && account.numberAssociations() + 1 > tokensConfig.maxPerAccount()),
```
Comment (MrValioBg) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenCreateValidator.java:195]: Is this change needed? It seems like it only adds `()`

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/EntitiesConfig.java [hunk 5]
```diff
@@ -24,4 +24,5 @@
 public record EntitiesConfig(
         @ConfigProperty(defaultValue = "3153600000") @NetworkProperty long maxLifetime,
         // @ConfigProperty(defaultValue = "FILE") Set<EntityType> systemDeletable
-        @ConfigProperty(defaultValue = "false") @NetworkProperty boolean limitTokenAssociations) {}
+        @ConfigProperty(defaultValue = "false") @NetworkProperty boolean limitTokenAssociations,
+        @ConfigProperty(defaultValue = "true") @NetworkProperty boolean unlimitedAutoAssociations) {}
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/EntitiesConfig.java:28]: Should this be `unlimitedAutoAssociationsEnabled` or `allowUnlimitedAutoAssociations` for clarity

Comment (Neeharika-Sompalli) [inline hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/EntitiesConfig.java:28]: Also this should be in `TokensConfig`

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CryptoCreateValidator.java [hunk 6]
```diff
@@ -98,7 +98,7 @@ public boolean tooManyAutoAssociations(
             @NonNull final LedgerConfig ledgerConfig,
             @NonNull final EntitiesConfig entitiesConfig,
             @NonNull final TokensConfig tokensConfig) {
-        return n > ledgerConfig.maxAutoAssociations()
-                || (entitiesConfig.limitTokenAssociations() && n > tokensConfig.maxPerAccount());
+        return (entitiesConfig.limitTokenAssociations() && n > tokensConfig.maxPerAccount())
+                || (!entitiesConfig.unlimitedAutoAssociations() && n > ledgerConfig.maxAutoAssociations());
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CryptoCreateValidator.java:102]: Why do we need to check `!entitiesConfig.unlimitedAutoAssociations()` this here ? Even if the configuration is enabled, we should not be able to set a higher number. That configuration is just allowed for `-1` to be set in `maxAutoAssociations` right ?

Comment (stoyanov-st) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CryptoCreateValidator.java:102]: From my understanding we should not limit the automatic associations if the configuration is enabled.
Or is it only if the value is -1 we have unlimited auto associations and we keep limiting the positive values e.g. 5000 is valid but 5001 - invalid?
@Neeharika-Sompalli @MiroslavGatsanoga  Can you provide more info on this?

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/EntitiesConfig.java (hunk 5 above)
Comment (stoyanov-st) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/EntitiesConfig.java:28]: I will change it to `allowUnlimitedAutoAssociations`

Comment (netopyr) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/EntitiesConfig.java:28]: I think it should be `unlimitedAutoAssociationsEnabled`, because we use this pattern more frequently.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CryptoCreateValidator.java (hunk 6 above)
Comment (netopyr) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CryptoCreateValidator.java:102]: I agree. If we allow unlimited auto associations, it is hard to explain why setting a concrete number should be limited. But this changes the current behavior and therefore we have to double-check (and eventually update the HIP).

Comment (MiroslavGatsanoga) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/CryptoCreateValidator.java:102]: > Or is it only if the value is -1 we have unlimited auto associations and we keep limiting the positive values e.g. 5000 is valid but 5001 - invalid
Yes, that should be the intended behaviour. Having the feature flag enabled should allow `maxAutoAssociations` value for any account to be set to -1.
If an account has `maxAutoAssociations = -1` then we have unlimited auto associations for this account.
